Tari script proposal v2

CjS77 opened the pull request:
This PR updates the TariScript proposal, fixing the vulnerability of the
data commitments by introducing the script key to the scheme.

CjS77 committed:
Tari script proposal v2

This PR updates the TariScript proposal, fixing the vulnerability of the
data commitments by introducing the script key to the scheme.

stringhandler commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -164,12 +167,20 @@ Therefore, it's imperative that the UTXO creator sign the script.
+The second is an addition of several pieces of data to the transaction kernel.
+The third is the addition of a new section in the transaction and block definition to hold the script data.
+
+Finally, the balance ans signature consensus rules must be updated to account for these changes.
```
```suggestion
Finally, the balance and signature consensus rules must be updated to account for these changes.
```

stringhandler commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
+$$
+\begin{align}
+\sigma_a &= \mathrm{H}(s_a) \\\\
+L_a &= k_{a'}. G \\\\
```
What is `a'` again?

CjS77 replied:
There's a notation table at the end. It represents the script key

stringhandler commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
-   & \sum(\mathrm{Inputs}) - \sum(\mathrm{Outputs}) - \sum(\mathrm{fee}_i.G)  \\\\
-  =& \sum\hat{C_i} - \sum\hat{C_j} - \sum(\mathrm{fee}_i.G)  \\\\
-  =& \sum(C_i + \mathrm{H}(C_i \Vert s_i).G) - \sum(C_j + \mathrm{H}(C_j \Vert s_j).G) - \sum(\mathrm{fee}.G)
+  & \sum(\mathrm{Inputs}) - \sum(\mathrm{Outputs}) - \sum(f_i.G)   \\\\
```
Is this meant to be `H`?
```suggestion
  & \sum(\mathrm{Inputs}) - \sum(\mathrm{Outputs}) - \sum(f_i.H)   \\\\
```

stringhandler reviewed (commented):
Some minor nits so long. Some examples with small integer private keys might help for understanding

delta1 commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
+  s \Rightarrow s^* = s - k_{a'}\sigma_a.e + k_{b'}\sigma_b.e
+$$
+
+To do this, Bob must know the value of \\( k_{a'} \\) which requires him to find the discrete log of \\( L_a \\), an infeasable axercise.
```
minor typo: `s/infeasable axercise/infeasible exercise/`

SWvheerden commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
+
+$$
+\begin{align}
+\sigma_a &= \mathrm{H}(s_a) \\\\
```
Is s_a here the script or the hash of the locking script? in line 172 `s = H(script)`

CjS77 replied:
s_a is the serialised script

SWvheerden commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
+### Script section
 
-The overall and block balance checks must also be modified to use \\( \hat{C} \\) rather than _C_.
+A new section in the Mimblewimble aggregate body is a set of scripts, and their associated data, for every input in the transaction.
```
I dont think aggregated body is a term that is standard of mimble wimble. Perhaps expand quickly what it is. Reading that as it is, to  me would be confusing if I did not know our code base.

SWvheerden commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
+The Mimblewimble transaction balance must be modified to take the script hash commitments into account. This is fairly straightforward.
 
-$$  
+Given a set of inputs, outputs, the transaction fee, \\( f \\), and the transaction offset, \\( \delta \\):
```
` \\( \delta \\):`
is never used?

CjS77 replied:
It's included now

SWvheerden commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
-the same but still manipulate specific outputs.
+Let's check whether the scripts are suitably committed on the blockchain, whether there's any malleability from miners or transaction participants and whether a new syncing node can verify the correctness of the chain.
+
+### Malleability
```
I think here is another malleability attack here. 

While Bob might not be able to replace ` k_{a'}\sigma_a.G`. He can swop it out. 
So he creates a new tx, with `hat{C} &= k_i.G + k_{b'}\sigma_b.G + v_b.H `
But v_b has very little value. So when he published the combined tx. He swaps `k_{b'}\sigma_b.G` with `k_{a'}\sigma_a.G`.

With the current  scheme no one should be able to detect that happend.

CjS77 replied:
Bob doesn't know the value of `k_{a'}\sigma_a.G` so I don't see how he can swap them out?

Even if it were possible, the range proof must be generated with `k_a + k_a'σ_a` so any modification to the commitment invalidates the range proof; and to recreate a valid proof you need to know both secrets.

SWvheerden reviewed (commented):
Looks good so far. I think we are progressing forward.

sdbondi commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
 $$
 
-If the accounting is correct, all values will cancel
+Here the \\( \pm \\) operator is used to simply the notation and indicates the addition operation when the following term is associated with an input and subtraction when the term related to an output.
```
```suggestion
Here the \\( \pm \\) operator is used to simplify the notation and indicates the addition operation when the following term is associated with an input and subtraction when the term related to an output.
```

CjS77 committed:
Further edits and refinements

CjS77 committed:
Add some more smart contract examples

Add Hash lock and HTLC, with examples. This entails defining the opcodes
that are needed to execute the script.

delta1 commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -441,54 +453,241 @@ The main properties of Tari script are
+with `ContextFailure`. It MUST reject the transaction if it fails with `Failure`.
+
+Not all context-dependent failures will fail with `ContextFailure`. For example, `CheckHeight(height) ZeroVerify` will
+fail with `Faiure` even though this script might pass in future. This is because the `CheckHeight(height)` does not have
```
s/`Faiure`/`Failure`/

delta1 commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -441,54 +453,241 @@ The main properties of Tari script are
+
+##### CheckSigVerify,
+
+Pop the public key and then the signature. A successful execution does not manipulate the stack an further.
```
s/`an`/`any`/

CjS77 committed:
WIP

CjS77 committed:
Merge branch 'development' into rfc_tari_script_v2

CjS77 committed:
More edits

dunnock commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -286,54 +487,247 @@ The proposal for Tari Script is straightforward. It is based on Bitcoin script a
+
+##### CheckSig,
+
+Pop the public key and then the signature. If the signature signs the script, push 0 to the stack, otherwise
```
Did not find it anywhere in the document might be simply missed it, but how would signature get into the stack in first place? Initially thought that supposed to be part of input data, but then thought it's recursion that first you need to have input data to be able to create signature..

CjS77 replied:
In this case the signature is in the input data. Input data is not part of the script itself, so it's possible to sign it without recursion issues.

But I do acknowledge that the definition of what's being signed is not well defined here and should be addressed.

dunnock commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -183,101 +210,275 @@ pub struct TransactionOutput {
+### Malleability
+
+Let's say a malicious party, Bob, can modify a transaction before it enters the blockchain.
+
```
Sorry if this was discussed already, did not find that before, but is it possible to perform kind of DOS attack on nodes supplying scripts which would be expensive to calculate and fail with `ContextFailure` at the end? Did not find limitations on number of times scripts executed and how fee is affecting repeats. Sample script:
`Blake256 (x 1016 times) CheckHeight(100_000)`

dunnock replied:
as a solution script might allow assertions which would lead to ContextFailure only at the beginning of the script.

CjS77 replied:
There are no loops, and there's a script size limit; so scripts cannot be that expensive.

CjS77 committed:
typos

CjS77 committed:
Merge branch 'development' into rfc_tari_script_v2

CjS77 committed:
Redraft Tari script

CjS77 committed:
Small edits

CjS77 commented:
Closed in favour of #2639
